autonat.go

```diff
@@ -75,199 +70,239 @@ func NewAutoNAT(ctx context.Context, h host.Host, getAddrs GetAddrs) AutoNAT {
-	as.status = s
-	switch s {
-	case NATStatusUnknown:
-		as.emitUnknown.Emit(event.EvtLocalRoutabilityUnknown{})
-	case NATStatusPublic:
-		as.emitPublic.Emit(event.EvtLocalRoutabilityPublic{})
-	case NATStatusPrivate:
-		as.emitPrivate.Emit(event.EvtLocalRoutabilityPrivate{})
-	}
+func (as *AmbientAutoNAT) emitStatus() {
+	status := as.status.Load().(autoNATResult)
+	as.emitReachabilityChanged.Emit(event.EvtLocalReachabilityChanged{Reachability: status.Reachability})
 }
 
+// PublicAddr returns the publicly connectable Multiaddr of this node if one is known.
 func (as *AmbientAutoNAT) PublicAddr() (ma.Multiaddr, error) {
-	as.mx.Lock()
-	defer as.mx.Unlock()
-
-	if as.status != NATStatusPublic {
+	s := as.status.Load().(autoNATResult)
+	if s.Reachability != network.ReachabilityPublic {
 		return nil, errors.New("NAT Status is not public")
 	}
 
-	return as.addr, nil
+	return s.address, nil
+}
+
+func ipInList(candidate ma.Multiaddr, list []ma.Multiaddr) bool {
+	candidateIP, _ := manet.ToIP(candidate)
+	for _, i := range list {
+		if ip, err := manet.ToIP(i); err == nil && ip.Equal(candidateIP) {
+			return true
+		}
+	}
+	return false
 }
 
 func (as *AmbientAutoNAT) background() {
 	// wait a bit for the node to come online and establish some connections
 	// before starting autodetection
-	select {
-	case <-time.After(AutoNATBootDelay):
-	case <-as.ctx.Done():
-		return
-	}
+	delay := AutoNATBootDelay
 
-	for {
-		as.autodetect()
+	var lastAddrUpdated time.Time
+	addrUpdatedChan := as.subAddrUpdated.Out()
+	defer as.subAddrUpdated.Close()
+	defer as.emitReachabilityChanged.Close()
 
-		delay := AutoNATRefreshInterval
-		if as.status == NATStatusUnknown {
-			delay = AutoNATRetryInterval
-		}
+	timer := time.NewTimer(delay)
+	timerRunning := true
 
+	for {
 		select {
-		case <-time.After(delay):
+		// new connection occured.
+		case conn := <-as.inboundConn:
+			localAddrs := as.host.Addrs()
+			ca := as.status.Load().(autoNATResult)
+			if ca.address != nil {
+				localAddrs = append(localAddrs, ca.address)
+			}
+			if !ipInList(conn.RemoteMultiaddr(), localAddrs) {
+				as.lastInbound = time.Now()
+			}
+
+		case <-addrUpdatedChan:
+			if !lastAddrUpdated.Add(time.Second).After(time.Now()) {
+				lastAddrUpdated = time.Now()
+				if as.confidence > 1 {
+					as.confidence--
+				}
+			}
+
+		// probe finished.
+		case result, ok := <-as.observations:
+			if !ok {
+				return
+			}
+			as.recordObservation(result)
+		case <-timer.C:
+			timerRunning = false
 		case <-as.ctx.Done():
 			return
 		}
-	}
-}
-
-func (as *AmbientAutoNAT) autodetect() {
-	peers := as.getPeers()
-
-	if len(peers) == 0 {
-		log.Debugf("skipping NAT auto detection; no autonat peers")
-		return
-	}
-
-	cli := NewAutoNATClient(as.host, as.getAddrs)
-	ctx, cancel := context.WithTimeout(as.ctx, AutoNATRequestTimeout)
-	defer cancel()
 
-	var result struct {
-		sync.Mutex
-		private int
-		public  int
-		pubaddr ma.Multiaddr
+		if timerRunning && !timer.Stop() {
```

Comment:
From the timer docs, just this should be fine:

```
	if !t.Stop() {
 		<-t.C
 	}
```

We can then get rid of the `timerRunning` variable.

Reply:
The reason for `timerRunning` is that if the timer fires above, then the channel will have already been drained. `t.Stop()` on a fired timer will return `false`, but in that code path we don't want to attempt to drain `t.C` again.

Reply:
Thanks for this.